### PR TITLE
Fix bug in TransformerMemcpy 

### LINKED ISSUE #1334: ssd-mobilenet model load fails with GPU backend
**Describe the bug**
Source of ssd-mobilenet: https://zenodo.org/record/3163026/files/ssd_mobilenet_v1_coco_2018_01_28.onnx (taken via MLPerf/inference https://github.com/mlperf/inference/tree/master/v0.5/classification_and_detection)

So this model correctly loads with CPU backend. It fails with the GPU backend. I tested it with the pypi build and a more recent master build with cuda 10.0

I try to load the model and the following error throws up:
 ```
   self._sess.load_model(path_or_bytes)
RuntimeError: [ONNXRuntimeError] : 9 : NOT_IMPLEMENTED : Using an input in multiple nodes on different devices is not supported currently. Input:Postprocessor/BatchMultiClassNonMaxSuppression/map/while/Slice/begin:0 is used by node Add__2519 (CUDAExecutionProvider) and node Postprocessor/BatchMultiClassNonMaxSuppression/map/while/Slice (CPUExecutionProvider).

```
Note that the model loads correctly with an onnxruntime build with backend of CPU.
Not sure if the issue is with the onnx file or the runtime or in the conversion. Any pointers / clarifications would help.

Also a bunch of these warnings popped up.
`[W:onnxruntime:Default, cuda_execution_provider.cc:1084 GetCapability] Fallback to CPU execution provider for Op type:
`
See attached file for whole log.[dump.log](https://github.com/microsoft/onnxruntime/files/3352451/dump.log)

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Ubuntu 18.04
- ONNX Runtime installed from (source or binary):
Source - master / binary pypi
- ONNX Runtime version:
0.4.0
- Python version:
3.6.8
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
7.4.0
- CUDA/cuDNN version:
10.0
- GPU model and memory:

**To Reproduce**
Describe steps/code to reproduce the behavior:
Download the model from the above mentioned link.
open python, 
import onnxruntime as rt
rt.InferenceSession('ssd_mobilenet_v1_coco_2018_01_28.onnx')

**Expected behavior**
Model loads correctly
**Screenshots**

**Additional context**
Add any other context about the problem here. If the issue is about a particular model, please share the model details as well to facilitate debugging.
So, in reference to https://github.com/microsoft/onnxruntime/issues/1229 
I downloaded the mentioned converted model from ssd_mobilenet_v2 as well and that fails similarly, although with CPU backend it did load correctly.

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: 

Fix a bug in TransformerMemcpy while processing initializers defined at an outer graph level when actually the initializer comes into scope in an inner subgraph level. Since the initializer(s) actually come into scope at an inner graph level, there is not enough information for the transformer to decide whether to introduce a copy node or not while processing the initializer(s) at an outer scope level. This oversight triggers a corner case failure and fails model load for certain models.

**Motivation and Context**
Resolve #1334

### CODE REVIEW

File: onnxruntime/test/framework/inference_session_test.cc [hunk 1]
```diff
@@ -1397,6 +1397,23 @@ TEST(InferenceSessionTests, TestParallelExecutionWithCudaProvider) {
   ASSERT_TRUE(!status.IsOK());
 }
 
+TEST(InferenceSessionTests, TestSSDMobileNet) {
```
Comment (hariharans29) [inline onnxruntime/test/framework/inference_session_test.cc]: Possibly there exists a better way to cover this change without testing it at an InferenceSession level and without having to checking-in a 28 MB model.

(For now keeping this here as a "test" placeholder)

File: onnxruntime/core/graph/graph_utils.h [hunk 2]
```diff
@@ -57,6 +57,11 @@ const ONNX_NAMESPACE::TensorProto* GetConstantInitializer(const Graph& graph, co
 void ReplaceInitializer(Graph& graph, const std::string& original_name, const ONNX_NAMESPACE::TensorProto& initializer,
                         bool check_outer_scope = true);
 
+/** Returns a pointer to the corresponding TensorProto for a name if it is an initializer. 
+@param check_outer_scope If true and the graph is a subgraph, check parent graph/s for 'name' if not found in 'graph'.
+*/
+const onnx::TensorProto* ReturnInitializedValue(const Graph& graph, const std::string& name, bool check_outer_scope);
+
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph_utils.h]: Maybe just put this locally in the memcpy transformer for now as I'm not aware of any other use case for it.

File: onnxruntime/test/framework/inference_session_test.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/test/framework/inference_session_test.cc]: There's an existing test where a graph is manually created with an initializer at outer scope. Perhaps it could be used as a base. ConstantFoldingSubgraph in onnxruntime\test\optimizer\graph_transform_test.cc

---
In reply to: [303690070](https://github.com/microsoft/onnxruntime/pull/1413#discussion_r303690070) [](ancestors = 303690070)

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 3]
```diff
@@ -150,15 +153,19 @@ bool TransformerMemcpyImpl::ModifyGraph(const KernelRegistryManager& kernel_regi
...
     const KernelCreateInfo* kci = nullptr;
     kernel_registries.SearchKernelRegistry(node, &kci);
 
     auto status = onnxruntime::Node::ForEachWithIndex(node.InputDefs(),
-                                                      [this, &kci](const onnxruntime::NodeArg& arg, size_t index) {
+                                                      [this, &kci, &initializers_scoped_to_current_graph_level](const onnxruntime::NodeArg& arg, size_t index) {
+                                                        // check if this NodeArg corresponds to an initializer defined in current graph or outer graph level 
+                                                        const auto* initializer_tensor_proto = graph_utils::ReturnInitializedValue(graph_, arg.Name(), true);
+                                                        if (initializer_tensor_proto != nullptr)
+                                                          initializers_scoped_to_current_graph_level[arg.Name()] = initializer_tensor_proto;
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: >initializers_scoped_to_current_graph_level [](start = 58, length = 42)

name seems slightly confusing when you can match initializers from outer scope. would something like 'initializers_consumed' be better?

File: onnxruntime/core/graph/graph_utils.h (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/graph/graph_utils.h]: Well is there value in introducing this to graph_utils even if there is no immediate use-case ? Moving it locally to MemCpy seems rather unnecessary especially since it does align with the overall theme of graph_utils....

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Definitely sounds better. Will change it.

File: onnxruntime/test/framework/inference_session_test.cc (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/test/framework/inference_session_test.cc]: Thanks, will create a test based on that one.

File: onnxruntime/core/graph/graph_utils.h (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph_utils.h]: It aligns with the theme but if it's only used by one thing it adds clutter. Up to you. If you leave it here maybe call it GetInitializer given it is the same as GetConstantInitializer sans the constant check. 

---
In reply to: [303695110](https://github.com/microsoft/onnxruntime/pull/1413#discussion_r303695110) [](ancestors = 303695110)

Comment (hariharans29) [reply onnxruntime/core/graph/graph_utils.h]: I d prefer to keep this in graph_utils for now. I made the change to the name of the method as suggested. Thanks!

**Comment by skottmckay** [comment]
>                                                       });

Please limit lines to 120 chars. If you move the first parameter to ForEachWithIndex to the next line clang-format should move the whole block in. [https://github.com/pharring/EditorGuidelines#getting-started](https://github.com/pharring/EditorGuidelines#getting-started) is very helpful as it displays a vertical line in the editor.

---
Refers to: onnxruntime/core/optimizer/transformer_memcpy.cc:181 in 379a88b. [](commit_id = 379a88b352332786fe7d47f0ed36f54e293fc2d5, deletion_comment = False)

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 4]
```diff
@@ -274,22 +288,37 @@ static const onnxruntime::NodeArg* FindNodeArg(const NodeArgSetType& def_set, co
...
       *(new_tensor_proto.mutable_name()) = new_def_name;
+
+      // We make a copy of the initializer that is to be consumed by the provider Node so that
+      // session state initializer can copy it over to the provider device during its operation
+      LOGS(logger_, INFO) << "Making a copy of the initializer: " << name << " and the copy's name is: " << new_def_name;
+
+      if (graph_.IsSubgraph()) {
+          // TODO: The copy being made is possibly redundant 
+          // When multiple subraphs consume the same initializer as an implicit input,
+          // mulitple copies of the initializer will be made into the provider device
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: >mulitple [](start = 13, length = 8)

multiple
[https://ewsoftware.github.io/VSSpellChecker/html/027d2fbc-7bfb-4dc3-b4f5-85f95fcf7629.htm](https://ewsoftware.github.io/VSSpellChecker/html/027d2fbc-7bfb-4dc3-b4f5-85f95fcf7629.htm) is very helpful

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 5]
```diff
@@ -322,6 +351,8 @@ void TransformerMemcpyImpl::ProcessInitializers(const KernelRegistryManager& ker
 
     p_node->ReplaceDefs(dup_replacements);
   }
+
+  return !replacements.empty();
 }
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: A comment saying this indicates the graph was modified would be helpful for someone new to the code

File: onnxruntime/test/framework/memcpy_transformer_test.cc [hunk 6]
```diff
@@ -177,6 +178,93 @@ TEST(TransformerTest, MemcpyTransformerTestCudaFirst) {
...
+  ASSERT_TRUE(status.IsOK()) << status.ErrorMessage();
+
+ // main graph continued
+  auto& if_node = graph.AddNode("node3", "If", "cpu operator2", ArgMap{&i1_def}, ArgMap{&o1_def, &o2_def});
+  if_node.AddAttribute("then_branch", {subgraph.ToGraphProto()});
+  if_node.AddAttribute("else_branch", {subgraph.ToGraphProto()});
+
+  onnxruntime::Graph* subgraph_1 = if_node.GetMutableGraphAttribute("then_branch");
+  for (auto& node : subgraph_1->Nodes()) {
+    if (node.Name() == "node2") {
```
Comment (skottmckay) [inline onnxruntime/test/framework/memcpy_transformer_test.cc]: Some comments here saying what you're doing and why would help someone new understand this test case.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Fixed.

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 7]
```diff
@@ -109,14 +116,17 @@ This transformer does not currently optimize copies between, e.g., two different
 
 bool TransformerMemcpyImpl::ModifyGraph(const KernelRegistryManager& kernel_registries) {
   bool modified = false;
+  InitializedTensorSet initializers_consumed;
   // find defs that require copy
   for (auto& node : graph_.Nodes()) {
     //don't need to do node placement here now, onnxruntime will do it according to registered kernels.
-    ProcessDefs(node, kernel_registries);
+    //as we process the defs for the nodes in the vurrent graph level, collect the initializers scoped to the current graph level alone
```
Comment (linkerzhang) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: typo: vurrent->current

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 8]
```diff
@@ -274,22 +290,37 @@ static const onnxruntime::NodeArg* FindNodeArg(const NodeArgSetType& def_set, co
...
     const onnxruntime::NodeArg* non_provider_def = FindNodeArg(non_provider_input_defs_, name);
     if (provider_def != nullptr && non_provider_def != nullptr) {
       std::string new_def_name = graph_.GenerateNodeArgName(name);
       auto& new_def = graph_.GetOrCreateNodeArg(new_def_name, provider_def->TypeAsProto());
 
-      const TensorProto* tensor_proto = nullptr;
-      bool found = graph_.GetInitializedTensor(name, tensor_proto);
-      ORT_ENFORCE(found, "Failed to get initialized tensor ", name);
+      const TensorProto* tensor_proto = pair.second;
+      ORT_ENFORCE(tensor_proto != nullptr, "Failed to get initialized tensor ", pair.first);
```
Comment (linkerzhang) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: this check is not needed. it will never be nullptr given the logic in ProcessDefs

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 9]
```diff
@@ -274,22 +290,37 @@ static const onnxruntime::NodeArg* FindNodeArg(const NodeArgSetType& def_set, co
...
 
       TensorProto new_tensor_proto = *tensor_proto;
       *(new_tensor_proto.mutable_name()) = new_def_name;
+
+      // We make a copy of the initializer that is to be consumed by the provider Node so that
+      // session state initializer can copy it over to the provider device during its operation
+      LOGS(logger_, INFO) << "Making a copy of the initializer: " << name << " and the copy's name is: " << new_def_name;
+
+      if (graph_.IsSubgraph()) {
```
Comment (linkerzhang) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: this is not needed. Adding a TODO or NOTE comment here is good enough.

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 10]
```diff
@@ -109,14 +116,17 @@ This transformer does not currently optimize copies between, e.g., two different
 
 bool TransformerMemcpyImpl::ModifyGraph(const KernelRegistryManager& kernel_registries) {
   bool modified = false;
+  InitializedTensorSet initializers_consumed;
   // find defs that require copy
   for (auto& node : graph_.Nodes()) {
     //don't need to do node placement here now, onnxruntime will do it according to registered kernels.
```
Comment (linkerzhang) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: this comment is confusing. I think it can be removed.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 7 above)
Comment (linkerzhang) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: and per your updated logic, this is not only collecting initializers in current graph level, but also its parent graph.

File: onnxruntime/core/optimizer/transformer_memcpy.cc [hunk 11]
```diff
@@ -4,24 +4,29 @@
...
 namespace onnxruntime {
 
 // implements MemCpy node insertion in graph transform
 // note that GraphTransformer::Apply() is supposed to be stateless, so this cannot derive from GraphTranformer
 class TransformerMemcpyImpl {
  public:
-  TransformerMemcpyImpl(onnxruntime::Graph& graph, const std::string& provider)
-      : graph_(graph), provider_(provider) {}
+  TransformerMemcpyImpl(onnxruntime::Graph& graph, const std::string& provider, const Logger& logger)
+      : graph_(graph), provider_(provider), logger_(logger) {}
```
Comment (linkerzhang) [inline onnxruntime/core/optimizer/transformer_memcpy.cc]: I don't see much value about adding logging info in this class.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 7 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: yes that is the intention...

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 8 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Removed this check. Thanks.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Thanks for this suggestion. It is pretty helpful.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 9 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: If you have multiple subgraphs using the same initializer from the main graph on a non-cpu device, each will copy the initializer. If that's a large weight there could be significant impact to the memory usage. It's difficult to handle better and may not come up enough to warrant that effort. Logging it provides an easy way to identify that this could be the source of unexpected memory usage. The log happens during loading so doesn't impact execution time. 

What's the reason for not wanting to log such a thing?

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 7 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: I think the point was the comment could be clarified to say that you're collection the initializers consumed at the current graph level. the initializers themselves could come from the current graph or any ancestor, so 'initializers scoped to the current graph level' is potentially confusing.

Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: I see. I updated the comment. Thanks.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 10 above)
Comment (linkerzhang) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: I mean this line "//don't need to do node placement here now, onnxruntime will do it according to registered kernels." :).

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 9 above)
Comment (linkerzhang) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: I got the point of trying to get attention that some copy may not be needed. Fair. then adding a TODO or even create a task tracking it. 

I would like to not have this logging, which does not add much value but overhead. The overhead may not be much, but, i'm being pushed more sensitive for some real cheap models from our customers. Let's not add more. Thank you!

Comment (skottmckay) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Sorry one more question. Given the log is at INFO level and our default is WARNING, it won't add cost unless someone changes the log level. Given that, is the guideline to still not have it? 

And is there a different guideline re: logging for code that happens during model load/initialization (like this) vs what happens during model execution? 

Maybe this info should be documented in the coding guidelines so it's more widely known.

File: onnxruntime/core/graph/graph_utils.h (hunk 2 above)
Comment (kkaranasos) [reply onnxruntime/core/graph/graph_utils.h]: Can we combine the two methods to one? With a boolean that specifies whether we should constant/overridable check.
Seems confusing to have two methods, especially given it is quite nuanced for most people what it means to have a non-constant initializer.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 9 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Removed all logging (kept comments) per offline discussion

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 10 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Sure, will remove this comment.

File: onnxruntime/core/optimizer/transformer_memcpy.cc (hunk 11 above)
Comment (hariharans29) [reply onnxruntime/core/optimizer/transformer_memcpy.cc]: Removed all logging.

File: onnxruntime/core/graph/graph_utils.h (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/graph/graph_utils.h]: Will scope this to just tranformer_memcpy.cc for now if it is confusing.
